File: app/src/main/java/com/teamunemployment/lolanalytics/FrontPage/Tabs/AnalyseTab/AnalyseInteractor.java
package com.teamunemployment.lolanalytics.FrontPage.Tabs.AnalyseTab;

import com.teamunemployment.lolanalytics.FrontPage.Tabs.AnalyseTab.Model.AnalyseData;

import java.util.ArrayList;

/**
 * Created by Josiah Kendall
 */

public class AnalyseInteractor {

    public void RequestFilterList(int role, int champId, AnalyseTabContract.Presenter presenter) {
        ArrayList<AnalyseData> datas = new ArrayList<>();
        AnalyseData analyseData = new AnalyseData();
        presenter.SetFilterRequestResponse(removeMeAfterMocking());
    }

    public void RequestFilterList(int role, AnalysePresenter presenter) {
        presenter.SetFilterRequestResponse(removeMeAfterMocking());
    }

    /**
     * TODO remove me
     */
    private ArrayList<AnalyseData> removeMeAfterMocking() {
        AnalyseData analyseData1 = new AnalyseData();
        AnalyseData analyseData2 = new AnalyseData();
        AnalyseData analyseData3 = new AnalyseData();
        AnalyseData analyseData4 = new AnalyseData();
        AnalyseData analyseData5 = new AnalyseData();
        AnalyseData analyseData6 = new AnalyseData();

        analyseData1.setRecentChange(-3.1);
        analyseData2.setRecentChange(-2.3);
        analyseData3.setRecentChange(3.3);

        analyseData1.setTitle("Early Game");
        analyseData2.setTitle("Mid Game");
        analyseData3.setTitle("Late Game");

        analyseData1.setHeroPercentTotal(45);
        analyseData2.setHeroPercentTotal(44);
        analyseData3.setHeroPercentTotal(47);

        analyseData1.setEnemyPercentTotal(55);
        analyseData2.setEnemyPercentTotal(56);
        analyseData3.setEnemyPercentTotal(53);

        ArrayList<AnalyseData> analyseDatalist = new ArrayList<AnalyseData>();
        analyseDatalist.add(analyseData1);
        analyseDatalist.add(analyseData2);
        analyseDatalist.add(analyseData3);
        return analyseDatalist;

    }
}

File: app/src/main/java/com/teamunemployment/lolanalytics/front_page/Tabs/StatTab/AnalyseInteractor.kt
package com.teamunemployment.lolanalytics.front_page.Tabs.StatTab

import co.metalab.asyncawait.async
import com.teamunemployment.lolanalytics.Utils.Network
import com.teamunemployment.lolanalytics.data.model.Result
import com.teamunemployment.lolanalytics.front_page.Tabs.StatTab.Model.StatList
import com.teamunemployment.lolanalytics.io.networking.RetrofitFactory
import ru.gildor.coroutines.retrofit.await

import java.util.ArrayList

/**
 * Created by Josiah Kendall
 */

class AnalyseInteractor(private val retrofitFactory: RetrofitFactory,
                         private val network: Network){
    private lateinit var playerAnalysisRemoteRepo : PlayerAnalysisRemoteRepo

    fun loadStatList(role : Int, summonerId: Long, games: Int, presenter: AnalysePresenter) {

        async {

            val url = await{ network.getUrl(summonerId) }
            playerAnalysisRemoteRepo = retrofitFactory.produceRetrofitInterface(PlayerAnalysisRemoteRepo::class.java, url)

            val result = playerAnalysisRemoteRepo.fetchStatList(summonerId, role, games).await()
            result.cache()
            presenter.handleListResult(result)
        }
    }

    fun loadStatList(role: Int, summonerId: Long, games: Int, champId: Int, presenter: AnalysePresenter) {
        async {
            val url = await{network.getUrl(summonerId)}
            playerAnalysisRemoteRepo = retrofitFactory.produceRetrofitInterface(PlayerAnalysisRemoteRepo::class.java, url)
            val result = playerAnalysisRemoteRepo.fetchStatList(summonerId, role, champId, games).await()
            result.cache()
            presenter.handleListResult(result)
        }
    }

    fun loadIndividualStat(url : String, viewHolder: AnalyseTabContract.CardView, presenter: AnalysePresenter) {
        async {
            val result = playerAnalysisRemoteRepo.fetchIndividualStat(url).await()
            result.cache()
            presenter.handleCardResult(result, viewHolder)
        }
    }

    suspend fun<T> Result<T>.cache() : Boolean {

        if (resultCode == 200) {
            cacheData(data)
            TODO("Implement caching for player analysis interactor")
            return true
        }
        return false
    }

    private fun<T> cacheData(data : T) : Boolean {
        when(data) {
            is StatList -> {
                // do stat cache
                return true
            }
            is ArrayList<*> -> {
                // get object 0 to figure out what it is?
                return true
            }
        }

        return false
    }
}
